sql-to-dbsp-compiler/lib/sqllib/src/casts.rs

```diff
@@ -315,91 +316,109 @@ cast_to_decimal!(u, usize);
 /////////// cast to decimalN
 
 #[inline]
-fn set_ps(value: Option<Decimal>, precision: u32, scale: i32) -> Option<Decimal> {
+fn set_ps(value: Option<SQLDecimal>, precision: u32, scale: i32) -> Option<SQLDecimal> {
     value.map(|v| cast_to_decimal_decimal(v, precision, scale))
 }
 
 #[inline]
-pub fn cast_to_decimalN_nullN(_value: Option<()>, _precision: u32, _scale: i32) -> Option<Decimal> {
+pub fn cast_to_decimalN_nullN(
+    _value: Option<()>,
+    _precision: u32,
+    _scale: i32,
+) -> Option<SQLDecimal> {
     None
 }
 
 #[inline]
-pub fn cast_to_decimalN_b(value: bool, precision: u32, scale: i32) -> Option<Decimal> {
+pub fn cast_to_decimalN_b(value: bool, precision: u32, scale: i32) -> Option<SQLDecimal> {
     let result = if value {
-        Some(Decimal::one())
+        Some(SQLDecimal::one())
     } else {
-        Some(Decimal::zero())
+        Some(SQLDecimal::zero())
     };
     set_ps(result, precision, scale)
 }
 
 #[inline]
-pub fn cast_to_decimalN_bN(value: Option<bool>, precision: u32, scale: i32) -> Option<Decimal> {
-    let result = value.map(|x| if x { Decimal::one() } else { Decimal::zero() });
+pub fn cast_to_decimalN_bN(value: Option<bool>, precision: u32, scale: i32) -> Option<SQLDecimal> {
+    let result = value.map(|x| {
+        if x {
+            SQLDecimal::one()
+        } else {
+            SQLDecimal::zero()
+        }
+    });
     set_ps(result, precision, scale)
 }
 
 #[inline]
-pub fn cast_to_decimalN_decimal(value: Decimal, precision: u32, scale: i32) -> Option<Decimal> {
+pub fn cast_to_decimalN_decimal(
+    value: SQLDecimal,
+    precision: u32,
+    scale: i32,
+) -> Option<SQLDecimal> {
     let result = Some(value);
     set_ps(result, precision, scale)
 }
 
 #[inline]
 pub fn cast_to_decimalN_decimalN(
-    value: Option<Decimal>,
+    value: Option<SQLDecimal>,
     precision: u32,
     scale: i32,
-) -> Option<Decimal> {
+) -> Option<SQLDecimal> {
     set_ps(value, precision, scale)
 }
 
 #[inline]
-pub fn cast_to_decimalN_d(value: F64, precision: u32, scale: i32) -> Option<Decimal> {
-    let result = Decimal::from_f64(value.into_inner());
+pub fn cast_to_decimalN_d(value: F64, precision: u32, scale: i32) -> Option<SQLDecimal> {
+    let result = SQLDecimal::from_f64(value.into_inner());
```

Comment:
Probably best to create a constructor that takes an f64 and a precision and scale directly